baselines/ppo2/layers.py

```diff
@@ -0,0 +1,260 @@
+
+            h, next_state = gru(input, mask, state, nh=nlstm)
+
+            num_layers = 2
+            num_hidden = 64
+            activation = tf.nn.relu
+            for i in range(num_layers):
+                h = fc(h, 'mlp_fc{}'.format(i), nh=num_hidden, init_scale=np.sqrt(2))
+                h = activation(h)
+
+            return h, next_state
+
+        return state, _network_fn
+
+    return network_fn
+
+
+def lstm(x, m, s, scope, nh, init_scale=1.0):
+    x = tf.layers.flatten(x)
+    nin = x.get_shape()[1]
+
+    with tf.variable_scope(scope):
+        wx = tf.get_variable("wx", [nin, nh * 4], initializer=ortho_init(init_scale))
+        wh = tf.get_variable("wh", [nh, nh * 4], initializer=ortho_init(init_scale))
+        b = tf.get_variable("b", [nh * 4], initializer=tf.constant_initializer(0.0))
+
+    m = tf.tile(tf.expand_dims(m, axis=-1), multiples=[1, nh])
+    c, h = tf.split(axis=1, num_or_size_splits=2, value=s)
+
+    c = c * (1 - m)
+    h = h * (1 - m)
+    z = tf.matmul(x, wx) + tf.matmul(h, wh) + b
+    i, f, o, u = tf.split(axis=1, num_or_size_splits=4, value=z)
+    i = tf.nn.sigmoid(i)
+    f = tf.nn.sigmoid(f)
+    o = tf.nn.sigmoid(o)
+    u = tf.tanh(u)
+    c = f * c + i * u
+    h = o * tf.tanh(c)
+    s = tf.concat(axis=1, values=[c, h])
+    return h, s
+
+
+def _ln(x, g, b, e=1e-5, axes=[1]):
+    u, s = tf.nn.moments(x, axes=axes, keep_dims=True)
+    x = (x - u) / tf.sqrt(s + e)
+    x = x * g + b
+    return x
+
+
+def lnlstm(x, m, s, scope, nh, init_scale=1.0):
+    x = tf.layers.flatten(x)
+    nin = x.get_shape()[1]
+
+    with tf.variable_scope(scope):
+        wx = tf.get_variable("wx", [nin, nh * 4], initializer=ortho_init(init_scale))
+        gx = tf.get_variable("gx", [nh * 4], initializer=tf.constant_initializer(1.0))
+        bx = tf.get_variable("bx", [nh * 4], initializer=tf.constant_initializer(0.0))
+
+        wh = tf.get_variable("wh", [nh, nh * 4], initializer=ortho_init(init_scale))
+        gh = tf.get_variable("gh", [nh * 4], initializer=tf.constant_initializer(1.0))
+        bh = tf.get_variable("bh", [nh * 4], initializer=tf.constant_initializer(0.0))
+
+        b = tf.get_variable("b", [nh * 4], initializer=tf.constant_initializer(0.0))
+
+        gc = tf.get_variable("gc", [nh], initializer=tf.constant_initializer(1.0))
+        bc = tf.get_variable("bc", [nh], initializer=tf.constant_initializer(0.0))
+
+    m = tf.tile(tf.expand_dims(m, axis=-1), multiples=[1, nh])
+    c, h = tf.split(axis=1, num_or_size_splits=2, value=s)
+
+    c = c * (1 - m)
+    h = h * (1 - m)
+    z = _ln(tf.matmul(x, wx), gx, bx) + _ln(tf.matmul(h, wh), gh, bh) + b
+    i, f, o, u = tf.split(axis=1, num_or_size_splits=4, value=z)
+    i = tf.nn.sigmoid(i)
+    f = tf.nn.sigmoid(f)
+    o = tf.nn.sigmoid(o)
+    u = tf.tanh(u)
+    c = f * c + i * u
+    h = o * tf.tanh(_ln(c, gc, bc))
+    s = tf.concat(axis=1, values=[c, h])
+
+    return h, s
+
+
+def gru(x, mask, state, nh, init_scale=-1.0):
```

Comment:
this is nitpicking, but I think gru should have the same call signature as lstm and lnlstm (and have `scope` argument)

Reply:
Okay 👍

Reply:
GRU is removed at 36aadd6
I would like to make GRU implementation PR If it is necessary.